### ISSUE TITLE
How set a section size in CTR-ACPKM cipher mode with CMS

### ISSUE THREAD

**Issue by phekd94** [openssl/openssl#16935]
We develop our engine that works with Russian gosts and try to realize CTR-ACPKM cipher mode (https://tools.ietf.org/id/draft-irtf-cfrg-re-keying-10.html#rfc.section.6.2.1). It works correct when encrypting data with the EVP API (NID is equal to NID_kuznyechik_ctr_acpkm).
But there is problem when I try use CMS.
```
EVP_CIPHER* cipher = EVP_get_cipherbynid(NID_kuznyechik_ctr_acpkm);
CMS_ContentInfo* cms = CMS_encrypt(..., inBio, cipher, 0); // calls EVP_CipherInit_ex() and EVP_EncryptUpdate()
```
I need to hard set section size value (N) (the number of bits that are processed with one section key before this key is transformed) due to OpenSSL doesn't set it via EVP_CIPHER_CTX_ctrl() with EVP_CTRL_KEY_MESH. But in gost this value must be different for various protocols (CMS, TLS1.2). Bellow I put section size value for various gosts:

- R 1323565.1.017-2018 (CTR-ACPKM gost)
  Magma        N = 1 Kbytes
  Kuznyechik  N = 4 Kbytes
- R 1323565.1.025-2019 (CMS gost)
  Magma        N = 8 Kbytes
  Kuznyechik  N = 256 Kbytes
- R 1323565.1.020-2018 (TLS1.2 gost)
  Magma        N = 1 Kbytes
  Kuznyechik  N = 4 Kbytes

How I can set section size when work with CMS? Or EVP_CIPHER_CTX_ctrl() with EVP_CTRL_KEY_MESH and correct section size should be called in CMS_encrypt()?

**Comment by beldmit** [comment]
For now it works. We have a default value for TLS and a separate value for CMS/PKCS#12. Both CMS and PKCS#12 read the ASN1 parameters, and the callback is involved. So the design is not the best possible, but fits the Russian GOST standards.

**Comment by buckythebear** [comment]
Looks like now all the GOST engines are forced to implement the behavior identical to the gost-engine for CMS_encrypt support in OpenSSL 3.0. Will the issue be solved better in OpenSSL 4.0 with providers? What's your vision, how will it be solved?

**Comment by beldmit** [comment]
> Looks like now all the GOST engines are forced to implement the behavior identical to the gost-engine for CMS_encrypt support in OpenSSL 3.0.

The behavior is prescribed by the standard. 



> Will the issue be solved better in OpenSSL 4.0 with providers? What's your vision, how will it be solved?

I don't see a problem now. In theory it is possible to add a control call providing a context (CMS/PKCS#12/TLS) but it still was not necessary.
